Remove legacy serialization and deserialization APIs

Lorak-mmk opened the pull request:
This PR finally removes both of the old APIs.
Focus in review should be on making sure that I didn't miss anything that shall be removed.

Removing old APIs makes it clear that some cleanups in scylla-cql shall be done.
Those will be described and done in a follow-up PR because this one is already big enough.

I'm delibarately not marking this as fixing https://github.com/scylladb/scylla-rust-driver/issues/1167 - the follow-up PR that deals with the consequences of API removal will do that.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [ ] ~~I added relevant tests for new features and bug fixes.~~
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [ ] ~~I have provided docstrings for the public items that I want to introduce.~~
- [x] I have adjusted the documentation in `./docs/source/`.
- [ ] I added appropriate `Fixes:` annotations to PR description.

wprzytula commented on scylla-cql/src/frame/value_tests.rs:
```diff
@@ -26,27 +25,24 @@ use std::sync::Arc;
 
-fn serialized_only_new<T: SerializeValue>(val: T, typ: ColumnType) -> Vec<u8> {
+// For now it is only used in a test that is also behind this feature flag,
+// so without it it throws a warning about being unised.
```
:pick: typo: `unised`

wprzytula commented on docs/source/conf.py:
```diff
@@ -64,6 +64,7 @@
 # Setup Sphinx
 def setup(sphinx):
     lexers['rust'] = RustLexer()
+    lexers['rust,ignore'] = RustLexer()
     lexers['toml'] = TOMLLexer()
```
:memo: I wasn't aware that there is such a place where we can configure how \`\`\`\<lang\> is parsed.

Lorak-mmk replied:
Me neither. In general I don't know much abut how Scylla documentation building works.

wprzytula commented on docs/source/migration-guides/0.15-deserialization.md:
```diff
@@ -1,20 +1,20 @@
-
-The old API and migration facilities will be removed in a future major release.
+In 0.15, a new deserialization API has been introduced. The new API improves type safety and performance of the old one, so it is highly recommended to switch to it. However, deserialization is an area of the API that users frequently interact with: deserialization traits appear in generic code and custom implementations have been written.
+In order to make migration easier, the driver 0.15 still offered the old API, which - while opt-in - can be very easily switched to after version upgrade.
```
:wrench: Leftover part.

wprzytula replied:
```suggestion
In order to make migration easier, the driver 0.15 still offered the old API.
```

wprzytula commented on docs/source/migration-guides/0.15-deserialization.md:
```diff
@@ -209,89 +209,18 @@ As mentioned in the Introduction section, the driver provides new procedural mac
-The impl generated by `FromUserType` expects UDT fields to be in the same order as the struct fields. Field names should be the same both in the UDT and in the struct. You can use the `enforce_order` attribute to achieve such behavior via the `DeserializeValue` trait.
+The impl generated by `FromUserType` expected UDT fields to be in the same order as the struct fields. Field names should be the same both in the UDT and in the struct. You can use the `enforce_order` attribute to achieve such behavior via the `DeserializeValue` trait.
 
 ### Adjusting custom impls of deserialization traits
 
 If you have a custom type with a hand-written `impl FromRow` or `impl FromCqlVal`, the best thing to do is to just write a new impl for `DeserializeRow` or `DeserializeValue` manually. Although it's technically possible to implement the new traits by using the existing implementation of the old ones, rolling out a new implementation will avoid performance problems related to the inefficient `CqlV ...
```
:wrench: No longer feasible.

wprzytula replied:
```suggestion
### Adjusting custom impls of deserialization traits

If you have a custom type with a hand-written `impl FromRow` or `impl FromCqlVal`, the best thing to do is to just write a new impl for `DeserializeRow` or `DeserializeValue` manually.
```

muzarski reviewed (commented):
Mainly some nits. I want to make sure that all removed tests for legacy API have their equivalent for new API (if applicable).

When it comes to the `value` module. I think that we should definitely make it a top-level module. Notice that deserialization tests are placed in `deserialize::value_tests`, while serialization tests are in `frame::value_tests`. So my proposition is to:
- move `value.rs` to top-level (`src/value.rs`)
- move `frame::value_tests.rs` to `serialize::value_tests.rs`

Lorak-mmk commented on docs/source/migration-guides/0.11-serialization.md:
````diff
@@ -6,7 +6,7 @@ When executing a statement through the CQL protocol, values for the bind markers
 Before 0.11, the driver couldn't do this kind of type checking. For example, in the case of non-batch queries, the only information about the user data it has is that it implements `ValueList` - defined as follows:
 
-```rust
+```rust,ignore
````
Yes, and this API no longer exists after this PR, so mdbook tests fail trying to build this code block.

Lorak-mmk commented on scylla-cql/src/frame/value_tests.rs:
```diff
@@ -1059,24 +1061,36 @@ fn serialized_values() {
-        fn serialize(&self, buf: &mut Vec<u8>) -> Result<(), ValueTooBig> {
+    impl SerializeValue for TooBigValue {
+        fn serialize<'b>(
+            &self,
+            _typ: &ColumnType,
+            writer: CellWriter<'b>,
+        ) -> Result<WrittenCellProof<'b>, SerializationError> {
             // serialize some
-            buf.put_i32(1);
+            writer.into_value_builder().append_bytes(&[1u8]);
 
             // then throw an error
-            Err(ValueTooBig)
+            Err(mk_ser_err::<Self>(
+                &ColumnType::Native(NativeType::TinyInt),
+                BuiltinTypeSerializationErrorKind::SizeOverflow,
+            ))
```
Right, nice catch

Lorak-mmk commented on scylla-cql/src/frame/value_tests.rs:
```diff
@@ -1328,57 +1304,15 @@ fn ref_value_list() {
 
-#[test]
-fn serialized_values_value_list() {
-    let mut ser_values = LegacySerializedValues::new();
-    ser_values.add_value(&1_i32).unwrap();
-    ser_values.add_value(&"qwertyuiop").unwrap();
-
-    let ser_ser_values: Cow<LegacySerializedValues> = ser_values.serialized().unwrap();
-    assert_matches!(ser_ser_values, Cow::Borrowed(_));
-
-    assert_eq!(&ser_values, ser_ser_values.as_ref());
+fn make_batch_value_iter<BV: BatchValues>(bv: &BV) -> BV::BatchValuesIter<'_> {
+    <BV as BatchValues>::batch_values_iter(bv)
 }
 
-#[test]
-fn cow_serialized_values_value_list() {
-    let cow_ser_values: Cow<LegacySerializedValues> = Cow::Owned(LegacySerializedValues::new());
-
-    let serialized: Cow<LegacySerializedValues> = cow_ser_values.serialized().unwrap();
-    assert_matches!(serialized, Cow::Borrowed(_));
-
-    assert_eq!(cow_ser_values.as_ref(), serialized.as_ref());
-}
```
Those tests test Cow-related functionality of the old API that is simply not present in the new one (and so the tests make no sense).
Old API had an "optimization" that `.serialized()` returned `Cow<...>`, so that you could not care if you already serialized the values or not, because serializing again is free (just borrows the data).

muzarski replied:
Makes sense, thanks for explanation

Lorak-mmk commented on scylla-cql/src/frame/response/cql_to_rust.rs:
```diff
@@ -957,102 +955,4 @@ mod tests {
-            MyRow::from_row(too_short_row),
-            Err(FromRowError::WrongRowSize {
-                expected: 3,
-                actual: 2
-            })
-        );
-
-        assert_eq!(
-            MyRow::from_row(too_large_row),
-            Err(FromRowError::WrongRowSize {
-                expected: 3,
-                actual: 4
-            })
-        );
-    }
-
-    // Enabling `expect_used` clippy lint,
-    // validates that `derive(FromRow)` macro definition does do not violates such rule under the hood.
-    // Could be removed after such rule will be applied for the whole crate.
-    // <https://rust-lang.github.io/rust-clippy/master/index.html#/expect_used>
-    #[deny(clippy::expect_used)]
-    #[test]
-    fn unnamed_struct_from_row() {
-        #[derive(FromRow)]
-        struct MyRow(i32, Option<String>, Option<Vec<i32>>);
-
-        let row = Row {
-            columns: vec![
-                Some(CqlValue::Int(16)),
-                None,
-                Some(CqlValue::Set(vec![CqlValue::Int(1), CqlValue::Int(2)])),
-            ],
-        };
-
-        let my_row: MyRow = MyRow::from_row(row).unwrap();
-
-        assert_eq!(my_row.0, 16);
-        assert_eq!(my_row.1, None);
-        assert_eq!(my_row.2, Some(vec![1, 2]));
-    }
```
Those tests test old value "deserialization" (converting from CqlValue to other types), and to a lesser degree row deserialization (converting from Row = Vec<Option<CqlValue>> to a concrete type).
New API has its own tests, which I believe to be extensive enough. See the `scylla-cql/src/deserialize/value_tests.rs` and `scylla-cql/src/deserialize/row_tests.rs`

Lorak-mmk replied:
I have no idea. @wprzytula ?

wprzytula replied:
No, there's no support for them. See respective [TODO](https://github.com/scylladb/scylla-rust-driver/blob/448502ca1db6228ab3d1d6108db78664e1dfddb6/scylla-macros/src/deserialize/mod.rs#L60).

wprzytula replied:
> Those tests test old value "deserialization" (converting from CqlValue to other types), and to a lesser degree row deserialization (converting from Row = Vec<Option> to a concrete type).

I believe that conversion of `CqlValue` into end types (using getters, e.g. `as_bigint()`) should be still well-tested, too, as it's a valid use case for, e.g., JS Driver. Is it already tested somewhere else?

Lorak-mmk replied:
In `serialize::value` module there are some tests for `CqlValue` serialization, but I don't see tests for `as_` method. I bielieve it is out of scope for this PR, so I'll open an issue about this.

Lorak-mmk commented:
- Addressed review comments
- Applied @wprzytula 's PR (https://github.com/Lorak-mmk/scylla-rust-driver/pull/4 )

Lorak-mmk commented:
Extracted value.rs and value_tests.rs from frame module.

Note: this was not a force push, only new commits were added, so there is no need to re-review previous commits if you already reviewed them.

muzarski reviewed (approved):
Docs need to be adjusted as well

Lorak-mmk commented:
> `no item named `response` in module `scylla_cql`

???!!!!!!

wprzytula commented on scylla-cql/src/serialize/value_tests.rs:
````diff
@@ -0,0 +1,2462 @@
+    fn _test_udt_bad_attributes_name_skip_name_checks_limitations_on_allow_missing() {}
+
+    /// ```
+    ///
+    /// #[derive(scylla_macros::SerializeValue)]
+    /// #[scylla(crate = scylla_cql, flavor = "enforce_order", skip_name_checks)]
+    /// struct TestUdt {
+    ///     a: i32,
+    ///     #[scylla(allow_missing)]
+    ///     b: bool,
+    ///     #[scylla(allow_missing)]
+    ///     c: String,
+    /// }
+    /// ```
+    fn _test_udt_good_attributes_name_skip_name_checks_limitations_on_allow_missing() {}
+
+    /// ```
+    /// #[derive(scylla_macros::SerializeValue)]
+    /// #[scylla(crate = scylla_cql)]
+    /// struct TestUdt {
+    ///     a: i32,
+    ///     #[scylla(allow_missing)]
+    ///     b: bool,
+    ///     c: String,
+    /// }
+    /// ```
+    fn _test_udt_unordered_flavour_no_limitations_on_allow_missing() {}
+
+    /// ```
+    /// #[derive(scylla_macros::SerializeValue)]
+    /// #[scylla(crate = scylla_cql)]
+    /// struct TestUdt {
+    ///     a: i32,
+    ///     #[scylla(default_when_null)]
+    ///     b: bool,
+    ///     c: String,
+    /// }
+    /// ```
+    fn _test_udt_default_when_null_is_accepted() {}
+}
````
:warning: ditto: never run doctests.

muzarski commented:
> > no item named `response` in module `scylla_cql`
> 
> ???!!!!!!

This is because of docstrings to `CqlDecimal` and `CqlVarint`: "e.g. conversion from [CqlValue](super::response::result::CqlValue)". which are placed in `value.rs`. The link to `CqlValue` got broken, since we moved `value` module - due to the usage of relative path (`super::*`). This could probably be adjusted to (`crate::*`), so no similar issues appear in the future.

Lorak-mmk commented:
> > > no item named `response` in module `scylla_cql`
> > 
> > 
> > ???!!!!!!
> 
> This is because of docstrings to `CqlDecimal` and `CqlVarint`: "e.g. conversion from CqlValue". which are placed in `value.rs`. The link to `CqlValue` got broken, since we moved `value` module - due to the usage of relative path (`super::*`). This could probably be adjusted to (`crate::*`), so no similar issues appear in the future.

Yep, I understood that a bit after posting my comment, I just didn't have time to fix the PR yey.

Lorak-mmk commented on scylla-cql/src/serialize/row_tests.rs:
````diff
@@ -0,0 +1,1064 @@
+    /// struct TestRow {}
+    /// ```
+    fn _test_struct_deserialization_name_check_skip_requires_enforce_order() {}
+
+    /// ```compile_fail
+    ///
+    /// #[derive(scylla_macros::SerializeRow)]
+    /// #[scylla(crate = scylla_cql, skip_name_checks)]
+    /// struct TestRow {
+    ///     #[scylla(rename = "b")]
+    ///     a: i32,
+    /// }
+    /// ```
+    fn _test_struct_deserialization_skip_name_check_conflicts_with_rename() {}
+
+    /// ```compile_fail
+    ///
+    /// #[derive(scylla_macros::SerializeRow)]
+    /// #[scylla(crate = scylla_cql)]
+    /// struct TestRow {
+    ///     #[scylla(rename = "b")]
+    ///     a: i32,
+    ///     b: String,
+    /// }
+    /// ```
+    fn _test_struct_deserialization_skip_rename_collision_with_field() {}
+
+    /// ```compile_fail
+    ///
+    /// #[derive(scylla_macros::SerializeRow)]
+    /// #[scylla(crate = scylla_cql)]
+    /// struct TestRow {
+    ///     #[scylla(rename = "c")]
+    ///     a: i32,
+    ///     #[scylla(rename = "c")]
+    ///     b: String,
+    /// }
+    /// ```
+    fn _test_struct_deserialization_rename_collision_with_another_rename() {}
+}
````
Huh, didn't know that. I'll drop the 2 commits that move doctests.

Lorak-mmk commented:
Fixed documentation and dropped the commit that moved the doctests.
One last thing we have to think about before 1.0 in this area: maybe CqlValue also should be moved to the new `scylla_cal::values` module? If so, that should be done in a follow-up PR.

Lorak-mmk commented on scylla-cql/src/serialize/batch_tests.rs:
```diff
@@ -0,0 +1,390 @@
+        }
+    }
+
+    // rustfmt wants to have each tuple inside a tuple in a separate line
+    // so we end up with 170 lines of tuples
+    // FIXME: Is there some cargo fmt flag to fix this?
```
I did not write this comment, it was just part of the moved code. I dislike the formatting a bit, but no matter if we want to change it or not I think it is out of scope for this PR.

Lorak-mmk commented:
> One last thing we have to think about before 1.0 in this area: maybe CqlValue also should be moved to the new scylla_cal::values module? If so, that should be done in a follow-up PR.

Now that I think of it I might as well do it here.
Why? I noticed that `frame/result.rs` contains value deserialization tests, which should be moved to `deserialize` module - similarly to what I did with `value_tests.rs` and `serialize` module.
@wprzytula @muzarski please confirm that `scylla-cql/src/value.rs` is an appropriate new home for `CqlValue`.

muzarski commented:
> > One last thing we have to think about before 1.0 in this area: maybe CqlValue also should be moved to the new scylla_cal::values module? If so, that should be done in a follow-up PR.
> 
> Now that I think of it I might as well do it here. Why? I noticed that `frame/result.rs` contains value deserialization tests, which should be moved to `deserialize` module - similarly to what I did with `value_tests.rs` and `serialize` module. @wprzytula @muzarski please confirm that `scylla-cql/src/value.rs` is an appropriate new home for `CqlValue`.

value.rs contains serializable/deserializable value types introduced in the driver (e.g. CqlVarint). It makes perfect sense to put CqlValue there as well.

Lorak-mmk commented:
Ok this PR is now +6k lines -11k lines and 37 commits. This is a signal that it got too big and should be split.
I'll leave this PR with the changes that remove old frameworks, and put all the resulting scylla_cql cleanups in a new PR.

Lorak-mmk commented:
Removed cleanup commits. Now this PR contains only the API removal.
I'll soon open new PR with the cleanups.

Lorak-mmk commented:
Opened https://github.com/scylladb/scylla-rust-driver/pull/1198

wprzytula commented on scylla/src/client/session_builder.rs:
```diff
@@ -590,10 +554,8 @@ impl<K: SessionBuilderKind> GenericSessionBuilder<K> {
-        &self,
-    ) -> Result<GenericSession<CurrentDeserializationApi>, NewSessionError> {
-        GenericSession::connect(self.config.clone()).await
+    pub async fn build(&self) -> Result<Session, NewSessionError> {
+        Session::connect(self.config.clone()).await
     }
```
:pushpin: Why is `build` borrowing instead of consuming the `SessionBuilder`? This makes no sense, as `SessionBuilder` implements `Clone`, so it could always be cloned before building to retain it. Now there is needless cloning for the usual case.

Lorak-mmk replied:
I can change that in the follow-up in order to not expand the scope of this PR
